### PR TITLE
feat: add crosschain properties

### PR DESCRIPTION
**Pull request by 0xDiscotech**
* refactor: allow for atomicity on mock messenger

* refactor: cross domain sender logic on mock messenger

* refactor: l2 to l2 caller and messenge sender logic

* chore: remove assumes for removed checks on mock messenger
----------------

On the mock messenger, I had to return immutable vars on cross domain sender function due to the same problem with the storage layout. The same happens with the destination chain value on `sendMessage`.

To run the files: `just test-halmos-all`

CLOSES FUZZ-32

### CODE REVIEW

File: packages/contracts-bedrock/test/properties/halmos/OptimismSuperchainERC20.t.sol [hunk 1]
```diff
@@ -112,37 +134,36 @@ contract SymTest_OptimismSuperchainERC20 is SymTest, AdvancedTest {
     function check_sendERC20ZeroCall(address _from, address _to, uint256 _chainId) public {
         /* Preconditions */
         vm.assume(_to != address(0));
-        vm.assume(_chainId != CURRENT_CHAIN_ID);
         vm.assume(_to != address(Predeploys.CROSS_L2_INBOX) && _to != address(MESSENGER));
 
-        uint256 _totalSupplyBefore = optimismSuperchainERC20.totalSupply();
-        uint256 _fromBalanceBefore = optimismSuperchainERC20.balanceOf(_from);
+        uint256 _totalSupplyBefore = sourceToken.totalSupply();
+        uint256 _fromBalanceBefore = sourceToken.balanceOf(_from);
```
Comment (simon-something) [inline packages/contracts-bedrock/test/properties/halmos/OptimismSuperchainERC20.t.sol:140]: nit> assert for `to` too

File: packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol [hunk 2]
```diff
@@ -2,22 +2,36 @@
...
+        DESTINATION_CHAIN_ID = _destinationChainId;
+        CUSTOM_CROSS_DOMAIN_SENDER = _customCrossDomainSender;
     }
 
-    function sendMessage(uint256 _destination, address _target, bytes calldata) external payable {
-        // TODO: Disable checks?
-        if (_destination == block.chainid) revert MessageDestinationSameChain();
-        if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
-        if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
+    function sendMessage(uint256 _destination, address, bytes calldata _message) external payable {
```
Comment (0xteddybear) [inline packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol:29]: shouldn't we keep the behaviour of reverting if the destination chain id is the same as the source chain id?

File: packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol [hunk 3]
```diff
@@ -28,23 +42,24 @@ contract MockL2ToL2Messenger is IL2ToL2CrossDomainMessenger {
         address _target,
         bytes calldata _message
     )
-        external
+        public
```
Comment (0xteddybear) [inline packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol]: why this change?

File: packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol (hunk 2 above)
Comment (0xDiscotech) [reply packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol:29]: We agreed that this won't be included on this PR, but could be on future ones after asking if we can extend the scope of the contracts to write invariant tests.

File: packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol (hunk 3 above)
Comment (0xDiscotech) [reply packages/contracts-bedrock/test/properties/halmos/MockL2ToL2Messenger.sol]: Don't know why, will undo

File: packages/contracts-bedrock/test/properties/halmos/OptimismSuperchainERC20.t.sol [hunk 4]
```diff
@@ -89,21 +114,18 @@ contract SymTest_OptimismSuperchainERC20 is SymTest, HalmosBase {
         vm.assume(_to != address(0));
         // Deploying a new messenger because of an issue of not being able to etch the storage layout of the mock
         // contract. So needed to a new one setting the symbolic immutable variable for the crossDomainSender.
-        vm.etch(address(MESSENGER), address(new MockL2ToL2Messenger(_crossDomainSender)).code);
+        // Used 0 address on source token so when the `soureToken` calls it if returns the symbolic `_crossDomainSender`
```
Comment (0xteddybear) [inline packages/contracts-bedrock/test/properties/halmos/OptimismSuperchainERC20.t.sol:117]: debugging this must've been the opposite of fun :grin:

File: packages/contracts-bedrock/test/properties/helpers/HalmosBase.sol [hunk 5]
```diff
@@ -5,6 +5,7 @@ import { Test } from "forge-std/Test.sol";
 
 contract HalmosBase is Test {
     uint256 internal constant CURRENT_CHAIN_ID = 1;
+    uint256 internal constant DESTINATION_CHAIN_ID = 2;
```
Comment (0xteddybear) [inline packages/contracts-bedrock/test/properties/helpers/HalmosBase.sol:8]: old variable isn't used anymore

**Review by 0xteddybear — COMMENTED** [review]
minor comments + properties should be marked as tested in `PROPERTIES.md`

**Comment by 0xDiscotech** [comment]
The symbolic tests will be written and proved using Kontrol instead
